design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,474 @@
+# Secrets Provider - Phase 2
+
+## Table of Contents
+- [Useful links](#useful-links)
+- [Background](#background)
+  * [Motivation](#motivation)
+  * [Requirement](#requirement)
+- [Solution](#solution)
+  * [Milestone 1: Deployment](#milestone-1--deployment)
+    + [Design](#design)
+    + [Flow](#flow)
+    + [Customer Experience](#customer-experience)
+    + [Code changes](#code-changes)
+    + [Fetching Relevant K8s Secrets](#fetching-relevant-k8s-secrets)
+      - [**Customer Experience**](#--customer-experience--)
+  * [Order of Deployment](#order-of-deployment)
+  * [Backwards compatibility](#backwards-compatibility)
+  * [Performance](#performance)
+  * [Affected Components](#affected-components)
+- [Security](#security)
+- [Test Plan](#test-plan)
+- [Logs](#logs)
+  * [Audit](#audit)
+- [Documentation](#documentation)
+- [Open questions](#open-questions)
+- [Implementation plan](#implementation-plan)
+
+## Glossary
+
+| **Term**                                                     | **Description**                                              |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| [Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/) | K8s entity that runs a pod for one-time operation. Once the pod exists, the Job terminates. |
+| [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) | K8s entity that ensures multiple replications of a pod are running all the time. If a pod terminates, another one will be created. |
+
+## Useful links
+
+| Name                | Link                                                         |
+| ------------------- | ------------------------------------------------------------ |
+| Phase 2 Feature Doc | https://github.com/cyberark/secrets-provider-for-k8s/issues/102 |
+
+## Background
+
+The Cyberark Secrets Provider for K8s enables customers to use secrets stored and managed in Conjur/DAP vault and consume them as K8s Secrets for their application containers.
+
+During Phase 1, an *init container* was deployed in the same Pod as the customer's applications to populate the K8s secrets defined in the pod application's manifest.
+
+### Motivation
+
+The current implementation of the Secrets Provider only supports one application and makes the deployment of the Secrets Provider deeply coupled to the customer's applications.
+
+### Requirements
+
+The solution should support the following capabilities:
+
+Milestone 1 *(current)*
+
+- Secrets Provider runs and deployed without affecting the app deployment or app life cycle
+- Secrets Provider can serve many apps
```

Comment:
Should we support secret provider serving number of namespaces ?

Reply:
@InbalZilberman ^ ?

Reply:
@sigalsax is it possible technically with one service-account?

Reply:
for a service account to serve multiple namespaces? yes if they have `clusterrole` but I dont think that this should be encouraged due to least privilege